### PR TITLE
ci: Add commitlint to ensure conventional commits

### PR DESCRIPTION
**Pull request by spetrosi**
Ensure that commits follow conventional commits format
Ensure that the PR title follows conventional commits format
Ensure that only one commit exists in the PR

### CODE REVIEW

**Comment by spetrosi** [comment]
@richm please review, the action that I am adding here is tested in https://github.com/linux-system-roles/mssql/pull/192. I am using [commitlint](https://github.com/conventional-changelog/commitlint), that resides in the conventional-changelog organization, so is provided by the same community that does the conventional commits initiative.

In addition to checking if commits meat the format, the action also checks that PR title meats the format, just for consistency. PR title is initially created from the commit, and if the commit was not in the format initially, the PR title might remain unformated.

The action also forces the PR to only have 1 commit, I think this will help us with collecting commits to changelog, wdyt? I think that squashing commits is a rule now too, just not forced by tooling. Hopefully, nobody will have problems with this.

File: playbooks/templates/.github/workflows/commitlint.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  merge_group:
+    branches:
+      - main
+    types:
+      - checks_requested
+jobs:
+  commit-checks:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (richm) [inline playbooks/templates/.github/workflows/commitlint.yml:13]:
```suggestion
      - uses: {{ gha_checkout_action }}
```

File: playbooks/templates/.github/workflows/commitlint.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        # PR, but rather to the merge commit inserted by the PR. So instead we
+        # have to get 'HEAD' from the PR event.
+        #
+        # One cannot use the number of commits
+        # (github.event.pull_request.commits) to find the start commit
+        # i.e. HEAD~N does not work, this breaks if there are merge commits.
+        run: |
+          echo "PR_HEAD=${{ github.event.pull_request.head.sha }}" >> $GITHUB_ENV
+          echo "PR_BASE=${{ github.event.pull_request.base.sha }}" >> $GITHUB_ENV
+
```
Comment (richm) [inline playbooks/templates/.github/workflows/commitlint.yml:48]:
```suggestion
{%- endraw +%}

```

File: playbooks/templates/.github/workflows/commitlint.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Run commitlint on commits
+        run: npx commitlint --from $PR_BASE --to $PR_HEAD --verbose
+
+      - name: Run commitlint on PR title
+        run: |
```
Comment (richm) [inline playbooks/templates/.github/workflows/commitlint.yml:53]:
```suggestion
        run: |
{%- raw %}
```

File: playbooks/templates/.github/workflows/commitlint.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          echo ${{ github.event.pull_request.title }} | npx commitlint --verbose
+
```
Comment (richm) [inline playbooks/templates/.github/workflows/commitlint.yml:55]:
```suggestion
{%- endraw +%}

```

File: playbooks/templates/.github/workflows/commitlint.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Fail if more than one commits exist in the PR
```
Comment (richm) [inline playbooks/templates/.github/workflows/commitlint.yml:56]:
```suggestion
      - name: Fail if more than one commits exist in the PR
{%- raw %}
```

File: playbooks/templates/.github/workflows/commitlint.yml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        if: ${{ github.event.pull_request.commits != 1 }}
```
Comment (richm) [inline playbooks/templates/.github/workflows/commitlint.yml:57]:
```suggestion
        if: ${{ github.event.pull_request.commits != 1 }}
{%- endraw +%}
```

**Comment by richm** [comment]
> @richm please review, the action that I am adding here is tested in [linux-system-roles/mssql#192](https://github.com/linux-system-roles/mssql/pull/192). I am using [commitlint](https://github.com/conventional-changelog/commitlint), that resides in the conventional-changelog organization, so is provided by the same community that does the conventional commits initiative.

Yes, looks good.  Just a few minor comments inline.
 
> In addition to checking if commits meat the format, the action also checks that PR title meats the format, just for consistency. PR title is initially created from the commit, and if the commit was not in the format initially, the PR title might remain unformated.

ok - is it ok if `[citest skip]` appears in the PR title?  Does this mean we can't have PR titles of the form `WIP - this is a work in progress`,  or is there a conventional commits way to do that?

> 
> The action also forces the PR to only have 1 commit, I think this will help us with collecting commits to changelog, wdyt? I think that squashing commits is a rule now too, just not forced by tooling. Hopefully, nobody will have problems with this.

I think some of our developers will have a problem with that.  It is common when reviewing PRs for developers to submit one or more commits for each PR review comment, then squash them all together when merging.  Not sure how to handle this situation.  Maybe we simply enforce that *all* commits have the head and body in the right format.  The developer is then responsible for either doing that, or squashing the commits into a single commit with the correct format before merging.  Note that there may be some situations where having multiple commits in a single PR, and having all of those commit messages go into the changelog, is desirable.

**Comment by spetrosi** [comment]
> > In addition to checking if commits meat the format, the action also checks that PR title meats the format, just for consistency. PR title is initially created from the commit, and if the commit was not in the format initially, the PR title might remain unformated.
> 
> ok - is it ok if `[citest skip]` appears in the PR title? Does this mean we can't have PR titles of the form `WIP - this is a work in progress`, or is there a conventional commits way to do that?

[citest skip] in the title works fine, tested in https://github.com/linux-system-roles/mssql/pull/192. I add an "edited" PR type trigger to default opened, synchronize, and reopened so that the check runs again after PR title is changed.

`WIP - this is a work in progress` - this will fail the check indeed, but that's fine to have the check failing until the PR is ready to be merged.

> > The action also forces the PR to only have 1 commit, I think this will help us with collecting commits to changelog, wdyt? I think that squashing commits is a rule now too, just not forced by tooling. Hopefully, nobody will have problems with this.
> 
> I think some of our developers will have a problem with that. It is common when reviewing PRs for developers to submit one or more commits for each PR review comment, then squash them all together when merging. Not sure how to handle this situation. Maybe we simply enforce that _all_ commits have the head and body in the right format. The developer is then responsible for either doing that, or squashing the commits into a single commit with the correct format before merging. Note that there may be some situations where having multiple commits in a single PR, and having all of those commit messages go into the changelog, is desirable.

One concern that I have here is that when you merge commits into one using `Squash and merge` in the GH UI, users are given a window where they edit their resulting commit themselves, and this resulting message becomes impossible to check because it gets merged right away. Typos in this stage might break the automation. The requirement to have a single commit solves this, but i understand that sometimes you need several commits in a PR.

Okay, I removed the condition to have only 1 commit, the action will check each commit to be in the right format.

**Comment by richm** [comment]
> using github UI to squash and merge allows editing the message, and that is not checked

Hmm - good point - if there really is no way to prevent a bad commit message using Squash&Merge, perhaps we should disable it, and have Rebase&Merge as the only merge option - then devs will need to squash manually and repush in order to squash.  Although, it would be nice to have a way for "superusers" to use Squash&Merge with the responsibility of creating a conforming commit message manually - but we already have a way to do that, it's called "go into the repo Settings and change the settings to allow you to do whatever you want before merging" . . .

**Comment by spetrosi** [comment]
Oh exactly, we should disallow squash merging and instead do Rebase&Merge on all repos. I tend to squash using CLI anyway because some time ago web UI was not very reliable, hopefully it is not a big deal for our devs.

**Comment by richm** [comment]
> Oh exactly, we should disallow squash merging and instead do Rebase&Merge on all repos.

Ok.  Please send an email to systemroles@lists.fedorahosted.org to explain the move to using Convention Commit message format, and explain that we will need to disable github UI Squash&Merge because of this.

> I tend to squash using CLI anyway because some time ago web UI was not very reliable, hopefully it is not a big deal for our devs.

+1 - the only time UI Squash&Merge is useful is when the developer is unable or unavailable to use the CLI, and it's easier for one of us admins to do it.  And, we can temporarily change the repo settings to allow it.

**Comment by spetrosi** [comment]
I sent an email, there are no replies to it, so I went on and disabled squash and merge on all repos. By the way, network had it disabled already.
Let's merge this PR now?
My plan is to write some simple scrip to reword commits from the previous tag to comply with conventional commits, so that we can automatically gather changelog for the next release.

**Comment by richm** [comment]
> I sent an email, there are no replies to it, so I went on and disabled squash and merge on all repos. By the way, network had it disabled already. Let's merge this PR now?

Yes.

> My plan is to write some simple scrip to reword commits from the previous tag to comply with conventional commits, so that we can automatically gather changelog for the next release.

ok, sounds good.

**Pull request merged**